### PR TITLE
Refactor Spatial Awareness System interfaces, replace system profile with observer profile(s)

### LINKED ISSUE #3171: vNext Task: Enable changing mesh display option at runtime
# Overview
As part of #2724, developers need a mechanism for changing the mesh display option of the spatial awareness system at runtime.

This functionality was present in HoloToolkit and should not require application developers to manage the mesh object materials.

# Requirements

* Application can dynamically switch between the mesh display options (visible, occlusion, none)

# Acceptance Criteria

- [ ] Application can dynamically switch the display option for the spatial mesh

### LINKED ISSUE #3226: NullReferenceExpection in BaseSpatialObserver.CleanupMeshObject on Hololens
**Does this affect the legacy HoloToolkit (master) or the Mixed Reality Toolkit (mrtk_release)?**
Mixed Reality Toolkit

**Describe the bug**
We added the spatial awareness system to visualize the room mesh. It works as expected on the hololens, but in the log I receive lots of NullReferenceExceptions in a very short timespan (see screenshot).

image

**To Reproduce**
The setup of the configuration profile looks like this:
image

So it is initially active and will be set programatically to hidden by using following code, but as far as I found out this switch is not relevant as the error occurs on the InvokeSurfaceDataReadyEvent (see stacktrace):

    private void SetSpatialMeshPulseVisible(bool visible) {
        bool isOpaque = MixedRealityToolkit.Instance.ActiveProfile.CameraProfile.IsOpaque;
        SpatialMeshDisplayOptions displayOptions = visible && !isOpaque ? SpatialMeshDisplayOptions.Visible : SpatialMeshDisplayOptions.None;
        MixedRealityToolkit.Instance.GetService<IMixedRealitySpatialAwarenessSystem>().MeshDisplayOption =
                displayOptions;
    }

**Unity Editor Version**
2018.2.17f1

**Mixed Reality Toolkit Release Version**
mrtk_development 49f5c4c22b66a7e2c212f15b982056dcb31ff4cd

### LINKED ISSUE #3326: spatial awareness interfaces changes (more feature focused)
## Overview

The Spatial Awareness System interface has too much of a burden for understanding the types of data a spatial observer may provide.

Similarly, putting a requirement on all observers to implement properties and methods, whether or not the type of data is provided, is not reasonable.

To resolve these issues, the IMixedRealitySpatialAwarenessSystem interface will be trimmed to provide only the core requirements for the system. Additionally, the IMixedRealitySpatialAwarenessObserver interface will be reduced to the essential components of an observer (ex: status, update interval, etc) with support for meshes, surface planes and other future types of data to be housed in data specific interfaces (ex: IMixedRealitySpatialAwarenessMeshObserver).

Enabling support for an observer that may return multiple types of data is critical as we should not require duplication of code or data synchronization issues in these scenarios.

## Requirements

- Spatial Awareness interfaces are to become more feature / task focused
- Existing concrete implementations are updated to use one or more of the interfaces
- Client code can access / control aspects of the observers as appropriate
- Existing client code (ex: demo scripts) should experience _minimal_ breaking changes

## Acceptance Criteria

- [ ] Interfaces define the minimal set of properties and methods to provide the feature
- [ ] Existing demo script continues to function as expected

### PR DESCRIPTION
**Pull request by davidkline-ms**
Overview
---
This change refactors the spatial awareness system to remove observer configuration settings from the system and the system profile. In fact, the spatial awareness system no longer needs / has a profile, instead each registered observer is individually configured.

As there were no supported planar surface observers, these settings have been removed from the spatial awareness system and will return in the future as part of an interface specifically to support spatial observers that return planar surface data natively.

Significant breaking changes include:
* IMixedRealitySpatialAwarenessSystem no longer contains configuration data for observers and now contains functions to Get/Suspend/Resume one or more observers
* IMixedRealitySpatialAwarenessSystemObserver and IMixedRealitySpatialAwarenessMeshObserver interfaces were added
* The DefaultMixedRealityToolkitConfigurationProfile has had the Spatial Awareness Profile removed
* The Windows Mixed Reality Spatial Awareness Mesh Observer (in the registered services profile) now has an associated profile.
* The class name for the Windows Mixed Reality observer has changed - If you have custom profiles referencing the observer, they will need to be updated.
* The MixedRealitySpatialAwarenessMeshHandler has been replaced with MixedRealitySpatialAwarenessObservationHandler - Please update any client code handing mesh events.

Fixes #3326 
Fixes #3171 
Fixes #3226

### CODE REVIEW

File: Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityControllerMappingProfile.cs [hunk 1]
```diff
@@ -36,6 +36,7 @@ private static void CollectControllerTypes()
             if (controllerMappingTypes == null)
             {
                 var tmp = new List<Type>();
+                // todo: not supported on uwp/.net
```
Comment (Cameron-Micka) [inline Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityControllerMappingProfile.cs]: Just double checking this is a capture TODO?

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        public float UpdateInterval => updateInterval;
+
+        #endregion IMixedRealitySpatialAwarenessObserver settings
+
+        #region IMixedRealitySpatialAwarenessMeshObserver settings
+
+        [PhysicsLayer]
+        [SerializeField]
+        [Tooltip("Physics layer on which to set observed meshes.")]
+        private int meshPhysicsLayer = 31;
```
Comment (Cameron-Micka) [inline Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs]: Nit: this is probably not a const expression, but would be a bit more consumable: https://docs.unity3d.com/ScriptReference/LayerMask.NameToLayer.html

File: Assets/MixedRealityToolkit/EventDatum/SpatialAwareness/MixedRealitySpatialAwarenessEventData.cs [hunk 3]
```diff
@@ -17,26 +18,54 @@ public class MixedRealitySpatialAwarenessEventData : GenericBaseEventData
...
         public int Id { get; private set; }
 
-        /// <summary>
-        /// <see cref="GameObject"/>, managed by the spatial awareness system, representing the data in this event.
-        /// </summary>
-        public GameObject GameObject { get; private set; }
+        ///// <summary>
+        ///// <see cref="SpatialAwarenessMeshObject"/>, managed by the spatial awareness system, representing the data in this event.
+        ///// </summary>
+        //public SpatialAwarenessMeshObject MeshObject{ get; private set; }
```
Comment (Cameron-Micka) [inline Assets/MixedRealityToolkit/EventDatum/SpatialAwareness/MixedRealitySpatialAwarenessEventData.cs]: Should we remove this?

File: Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityControllerMappingProfile.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityControllerMappingProfile.cs]: this is a note to fix building against the .net backend in a future pr

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/BaseSpatialAwarenessObject.cs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+        /// <inheritdoc />
+        public MeshFilter Filter { get; set; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        public virtual void CleanObject()
+        {
+            // todo: consider if this should be virtual, and what params it should contain
```
Comment (radicalad) [inline Assets/MixedRealityToolkit/Definitions/SpatialAwareness/BaseSpatialAwarenessObject.cs]: it is virtual?

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs (hunk 2 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs]: besides not being a const. the name might be changed by the user....

File: Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityControllerMappingProfile.cs (hunk 1 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Definitions/Devices/MixedRealityControllerMappingProfile.cs]: this is to remind me to fix building the .net backend. it is not related to this pr.

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+        [SerializeField]
+        [Tooltip("Level of detail, in triangles per cubic meter.\nIgnored unless LevelOfDetail is set to Custom.")]
+        private int trianglesPerCubicMeter = 0;
+
+        /// <summary>
+        /// The level of detail, in triangles per cubic meter, for the returned spatial mesh.
+        /// </summary>
+        /// <remarks>This value is ignored, unless <see cref="LevelOfDetail"/> is set to Custom.</remarks>
+        public int TrianglesPerCubicMeter => trianglesPerCubicMeter;
```
Comment (radicalad) [inline Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs]: do we want a setter here or is this something that can only be set in an inspector/profile?

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        public bool RecalculateNormals => recalculateNormals;
+
+        [SerializeField]
+        [Tooltip("How should spatial meshes be displayed?")]
+        private SpatialAwarenessMeshDisplayOptions displayOption = SpatialAwarenessMeshDisplayOptions.Visible;
+
+        /// <summary>
+        /// Indicates how the mesh subsystem is to display surface meshes within the application.
+        /// </summary>
+        public SpatialAwarenessMeshDisplayOptions DisplayOption => displayOption;
```
Comment (radicalad) [inline Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs]: A bunch of these look like they should have a setter.

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+        [SerializeField]
+        [Tooltip("Material to use when displaying observed meshes")]
+        private Material visibleMaterial = null;
+
+        /// <summary>
+        /// The material to be used when displaying observed meshes.
+        /// </summary>
+        public Material VisibleMaterial => visibleMaterial;
```
Comment (radicalad) [inline Assets/MixedRealityToolkit/Definitions/SpatialAwareness/MixedRealitySpatialAwarenessMeshObserverProfile.cs]: this too.

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+using Microsoft.MixedReality.Toolkit.Core.Devices;
+using Microsoft.MixedReality.Toolkit.Core.Services;
+using System;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Core.Definitions.SpatialAwarenessSystem
+{
+    public partial class SpatialAwarenessPlanarObject : BaseSpatialAwarenessObject
+    {
+        // todo
```
Comment (radicalad) [inline Assets/MixedRealityToolkit/Definitions/SpatialAwareness/SpatialAwarenessPlanarObject.cs]: remove

File: Assets/MixedRealityToolkit/Definitions/Utilities/ProfileMenuItemIndices.cs [hunk 9]
```diff
@@ -18,7 +18,9 @@ public enum CreateProfileMenuItemIndices
         Speech,
         BoundaryVisualization,
         ControllerVisualization,
-        SpatialAwareness,
+        SpatialAwareness,   // todo: remove
```
Comment (radicalad) [inline Assets/MixedRealityToolkit/Definitions/Utilities/ProfileMenuItemIndices.cs]: should we go ahead and remove it now?

File: Assets/MixedRealityToolkit/Definitions/SpatialAwareness/BaseSpatialAwarenessObject.cs (hunk 4 above)
Comment (davidkline-ms) [reply Assets/MixedRealityToolkit/Definitions/SpatialAwareness/BaseSpatialAwarenessObject.cs]: not needed at all... removed

**Pull request merged**